executor/docs/spec.md

```diff
@@ -0,0 +1,276 @@
+# Cook Executor
+
+The document outlines a specification for Cook's custom executor.
+
+It is a modular design allowing users to add functionality by writing reusable, single-responsibility processes. Declarative configuration would determine when then processes would be executed. An HTTP API exposed by the executor itself would enable communication between the executor and the tasks it starts.
+
+
+## Scheduler HTTP API
+
+The following enhancements to the existing Cook scheduler API will:
+
+1. Allow users to specify commands that run before, after, or during the primary command when creating jobs
+2. Allow users to query the exit code of a completed job
+
+### Launch a Job
+
+It should be possible to specify additional commands when launching a job. Additional commands should run either before, after, or during the primary command. Any failures (non-zero exit codes) will cause the entire job to fail. In the future, it may be desirable to have finer-grained control over how failure will be handled (for example, ignoring failures from asynchronous monitoring commands).
+
+Request body schema additions:
+
+| JSON Key        | Type                     | Description                              |
+| --------------- | ------------------------ | ---------------------------------------- |
+| before_commands | array of command objects | the commands to run before the real command |
+| after_commands  | array of command objects | the commands to run after the real command |
```

Comment:
Do the after commands run regardless of whether the real command succeeded? Is there a way to access the exit code?